File: scr/P205.cpp
#include <iostream>
#include <vector>
#include <map>
#include <chrono>

using namespace std;
using namespace std::chrono;

// Function to generate all possible outcomes of rolling dice
vector<int> rollDice(int numDice, int sides) {
    vector<int> outcomes;
    int maxOutcome = pow(sides, numDice);
    for (int i = 0; i < maxOutcome; i++) {
        int sum = 0, temp = i;
        for (int j = 0; j < numDice; j++) {
            sum += temp % sides + 1;
            temp /= sides;
        }
        outcomes.push_back(sum);
    }
    return outcomes;
}

// Function to calculate the probability distribution
map<int, double> calculateProbabilityDistribution(int numDice, int sides) {
    vector<int> outcomes = rollDice(numDice, sides);
    map<int, int> frequency;
    map<int, double> probabilityDistribution;

    for (int sum : outcomes) {
        frequency[sum]++;
    }

    int totalOutcomes = outcomes.size();
    for (auto& p : frequency) {
        probabilityDistribution[p.first] = (double)p.second / totalOutcomes;
    }

    return probabilityDistribution;
}

// Timer decorator equivalent in C++
double P205() {
    auto start = high_resolution_clock::now();

    // Calculate probability distribution for Peter (9 four-sided dice)
    map<int, double> peterDistribution = calculateProbabilityDistribution(9, 4);

    // Calculate probability distribution for Colin (6 six-sided dice)
    map<int, double> colinDistribution = calculateProbabilityDistribution(6, 6);

    double peterBeatsColinProbability = 0.0;

    for (auto& peter : peterDistribution) {
        double colinProbSum = 0.0;
        for (auto& colin : colinDistribution) {
            if (colin.first < peter.first) {
                colinProbSum += colin.second;
            }
        }
        peterBeatsColinProbability += peter.second * colinProbSum;
    }

    auto stop = high_resolution_clock::now();
    auto duration = duration_cast<microseconds>(stop - start);
    cout << "Execution time: " << duration.count() << " microseconds" << endl;

    return round(peterBeatsColinProbability * 10000000) / 10000000.0;
}

int main() {
    double result = P205();
    cout << "Probability that Pyramidal Peter beats Cubic Colin: " << result << endl;
    return 0;
}

File: scr/P205.R
# Function to calculate the probability distribution of dice rolls
calculateProbabilityDistribution <- function(numDice, sides) {
  outcomes <- expand.grid(replicate(numDice, 1:sides, simplify = FALSE))
  sums <- rowSums(outcomes)
  frequency <- table(sums)
  probabilityDistribution <- frequency / length(sums)
  return(probabilityDistribution)
}

# Calculate probability distribution for Peter (9 four-sided dice)
peterDistribution <- calculateProbabilityDistribution(9, 4)

# Calculate probability distribution for Colin (6 six-sided dice)
colinDistribution <- calculateProbabilityDistribution(6, 6)

# Function to calculate the probability that Peter's sum is greater than Colin's
peterBeatsColinProbability <- function(peterDistribution, colinDistribution) {
  probability <- 0
  for (peterSum in names(peterDistribution)) {
    colinProbSum <- sum(colinDistribution[names(colinDistribution) < peterSum])
    probability <- probability + peterDistribution[peterSum] * colinProbSum
  }
  return(probability)
}

# Measure execution time
start.time <- Sys.time()
result <- peterBeatsColinProbability(peterDistribution, colinDistribution)
end.time <- Sys.time()
execution.time <- end.time - start.time

# Output
cat("Probability that Pyramidal Peter beats Cubic Colin:", round(result, 7), "\n")
cat("Execution time:", execution.time, "seconds\n")
